File: boj/9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        for(int t = Integer.parseInt(br.readLine()); t-- > 0;) {
            String s = br.readLine();
            int len = s.length();
            int sum = 0, flag = 1;
            for(int i = 0; i < len; ++i) {
                if(s.charAt(i) == '(') {
                    ++sum;
                } else {
                    --sum;
                    if(sum < 0) {
                        flag = 0;
                        break;
                    }
                }
            }
            if(sum > 0) flag = 0;
            sb.append((flag == 1) ? "YES\n" : "NO\n");
        }
        System.out.println(sb);
    }
}

File: boj/9012.cpp
#include <iostream>

using namespace std;

int main(void) {
    ios_base::sync_with_stdio(false);
    int t; for (cin >> t; t--;) {
        string s; cin >> s;
        int len = s.length();
        int sum = 0, flag = 1;
        for (int i = 0; i < len; ++i) {
            if (s[i] == '(') {
                ++sum;
            }
            else {
                --sum;
                if (sum < 0) {
                    flag = 0;
                    break;
                }
            }
        }
        if (sum > 0) flag = 0;
        cout << (flag ? "YES\n" : "NO\n");
    }
    return 0;
}
